xarray/tests/test_backends.py

```diff
@@ -1519,6 +1519,23 @@ def test_encoding_unlimited_dims(self):
             assert actual.encoding["unlimited_dims"] == set("y")
             assert_equal(ds, actual)
 
+    def test_write_netcdf4_to_memory(self):
+        expected = xr.Dataset({"v": xr.DataArray(data=np.arange(10))})
+        buf = expected.to_netcdf(engine="netcdf4")
+        with open_dataset("dummy.nc", engine="netcdf4", memory=buf) as actual:
+            assert_equal(expected, actual)
+
+    def test_read_from_memory(self):
+        x = np.arange(10)
+        # The following doesn't create an actual file
+        ds = nc4.Dataset("dummy.nc", "w", memory=4096)
+        ds.createDimension("time", size=10)
+        ds.createVariable("x", np.int32, dimensions=("time",))
+        ds.variables["x"][:] = x
+        buf = ds.close()
+        with open_dataset("dummy.nc", engine="netcdf4", memory=buf) as ds:
+            assert all(ds.x == x)
+
```

Comment:
Re the test failures below in `test_engine`, I think you can remove the `pytest.raises` for the case without a filename now.